### ISSUE TITLE
[FEATURE] Reduce impact of checking if Lombok is disabled

### ISSUE THREAD

**Issue by dreis2211** [projectlombok/lombok#3361]
Hi,

this time no direct PR because I have no idea how to proceed with that. I've noticed that every compilation unit is checked on the fact if Lombok is disabled or not in `JavacTransformer`.

```
public void transform(long priority, Context context, List<JCCompilationUnit> compilationUnits, CleanupRegistry cleanup) {
	for (JCCompilationUnit unit : compilationUnits) {
		if (!Boolean.TRUE.equals(LombokConfiguration.read(ConfigurationKeys.LOMBOK_DISABLE, JavacAST.getAbsoluteFileLocation(unit)))) {
			// ... 
		}
	}
}
```

There are two things that pop out here.
- The purple highlight of reading the actual configuration. Which is currently at around ~1% (might be less when my other PRs are merged though)
- Getting the absolute file location of the compilation unit, which ultimately leads to the `stat64` syscall (for checking if it's a directory or not), which is at ~1.5% currently

<img width="860" alt="image" src="https://user-images.githubusercontent.com/6304496/221117700-bfa1dc9e-3749-4b2d-bd0c-ee6831cbfb5e.png">

Overall this accounts for 2-3% at the moment. Which is unfortunate, because in 99% of the cases you don't have Lombok disabled  I'd say.
Moreover, most of the time you only have no `lombok.config` files at all, maybe one, sometimes a few in your project. It seems like an overhead to check if there is a `lombok.config` alongside of every compilation unit.

I'm wondering if it would make sense to refactor the config loading overall and check for the `lombok.config` files already upfront and pass them to the transformer. Or given that `ConfigurationKeys.LOMBOK_DISABLE` is rarely used, doing a check on all these upfront if the key is present somewhere and if not skipping this check alltogether.

I really have no clue how to tackle this best. Maybe you have ideas, hence me sharing this.

Cheers,
Christoph

**Comment by Rawi01** [comment]
I agree that the lombok config system is not perfect and performs a huge amount of IO operations which leads to some overhead. Lombok already use caches to reduce that overhead. This is even worse for eclipse if you use a system where IO is slow due to AV software etc.

I also thougt about collection all config files in the beginning and reuse them. While this might be useful for `javac` I doubt that it helps in eclipse if we only compile one file. I never benchmarked this because I always hope that someday I have a perfect idea to solve all this properly.

To keep backward compatibility we can not change that it is allowed to add a single config file per package/folder. A breaking change is always a bit unpleasant. In case someone does not use the config system it can be disable by setting `lombok.disableConfig` but I guess that this will not disable the resolution of the absolute file location.

Without having checked, I think that the method you mentioned is simply the first one that uses the config system and all other config calls are more efficient because they end up using the cached values.

A few weeks ago I noticed that there are two methods to get the absolute file location and only uses the cache. I already patched that but never pushed that somewhere (stupid me).
